grabber/main.go

```diff
@@ -208,77 +217,103 @@ func checkAncestors(importer *backend.Backend, blockNumber int64, generations in
 	lastBlockUpdatedAt := int64(0)
+	t := time.NewTicker(sleep)
+	defer t.Stop()
 	for {
-		start := time.Now()
-		currentTime := time.Now()
-		currentBlock := getLatestBlockNumber(importer)
-		addresses := importer.GetActiveAdresses(lastUpdatedAt, updateContracts)
-		importer.Lgr.Info("updateAddresses:", zap.Time("lastUpdatedAt", lastUpdatedAt), zap.Int("Addresses in db", len(addresses)))
-		var jobs = make(chan *models.ActiveAddress, workersCount)
-		var wg sync.WaitGroup
-		for i := 0; i < int(workersCount); i++ {
-			wg.Add(1)
-			go worker(&wg, jobs, currentBlock, blockRangeLimit, importer)
-		}
-		for _, address := range addresses {
-			jobs <- address
+		select {
+		case <-t.C:
+			start := time.Now()
+			currentTime := time.Now()
+			currentBlock, err := importer.GetLatestBlockNumber()
+			if err != nil {
+				importer.Lgr.Error("Update Addresses: Failed to get latest block number", zap.Error(err))
+				continue
+			}
+			addresses := importer.GetActiveAdresses(lastUpdatedAt, updateContracts)
+			importer.Lgr.Info("Update Addresses: Starting", zap.Time("lastUpdatedAt", lastUpdatedAt), zap.Int("count", len(addresses)))
+			var jobs = make(chan *models.ActiveAddress, workersCount)
+			var wg sync.WaitGroup
+			for i := 0; i < int(workersCount); i++ {
+				wg.Add(1)
+				lgr := importer.Lgr.With(zap.Int("worker", i))
+				go worker(wg.Done, lgr, jobs, currentBlock, blockRangeLimit, importer)
+			}
+			for _, address := range addresses {
+				jobs <- address
+			}
+			close(jobs)
+			wg.Wait()
+			elapsed := time.Since(start)
+			importer.Lgr.Info("Update Addresses: Complete", zap.Bool("updateContracts", updateContracts),
+				zap.Duration("elapsed", elapsed), zap.Int64("block", lastBlockUpdatedAt))
+			lastBlockUpdatedAt = currentBlock
+			lastUpdatedAt = currentTime
 		}
-		close(jobs)
-		wg.Wait()
-		elapsed := time.Since(start)
-		importer.Lgr.Info("Performance measurement:", zap.Bool("updateContracts", updateContracts), zap.String("Updating all addresses took", elapsed.String()), zap.Int64("Current block", lastBlockUpdatedAt))
-		lastBlockUpdatedAt = currentBlock
-		lastUpdatedAt = currentTime
-		time.Sleep(sleep)
 	}
 }
 
-func worker(wg *sync.WaitGroup, jobs chan *models.ActiveAddress, currentBlock int64, blockRangeLimit uint64, importer *backend.Backend) {
+func worker(done func(), lgr *zap.Logger, jobs chan *models.ActiveAddress, currentBlock int64, blockRangeLimit uint64, importer *backend.Backend) {
+	defer done()
+	var errs []error
+	var updated int
 	for address := range jobs {
-		updateAddress(address, currentBlock, blockRangeLimit, importer)
+		err := updateAddress(address, currentBlock, blockRangeLimit, importer)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to update address %s: %v", address.Address, err))
+		} else {
+			updated++
+		}
+	}
+	if len(errs) > 0 {
+		lgr.Error("Update Addresses: worker done", zap.Int("updated", updated), zap.Int("failed", len(errs)), zap.Errors("errors", errs))
```

Comment:
Other fatals are just logged as failed attempts per address, and they are retried each time.